src/unix/thread.c

```diff
@@ -10,13 +10,25 @@ sysreturn gettid()
 
 sysreturn set_tid_address(int *a)
 {
+    /* man page says this always succeeds, but... */
+    if (!validate_user_memory(a, sizeof(int), true))
+        return set_syscall_error(current, EFAULT);
     current->clear_tid = a;
     return current->tid;
 }
 
 sysreturn arch_prctl(int code, unsigned long addr)
 {    
     thread_log(current, "arch_prctl: code 0x%x, addr 0x%lx", code, addr);
+    if ((code == ARCH_GET_FS || code == ARCH_GET_GS) &&
+        !validate_user_memory((void *)addr, sizeof(u64), true))
+        return set_syscall_error(current, EFAULT);
```

Comment:
Now that we have these checks, the code `if (!addr) return set_syscall_error(current, EINVAL);` should be removed from the ARCH_GET_FS and ARCH_GET_GS cases below.